openrag/utils/temporal.py

```diff
@@ -0,0 +1,286 @@
+"""
+Temporal utilities for query normalization and date extraction.
+Language-agnostic temporal expression extraction.
+"""
+
+import re
+from datetime import datetime, timedelta, timezone
+from typing import Optional, Dict, Tuple
+
+
+class TemporalQueryNormalizer:
+    """
+    Extracts temporal expressions from queries and normalizes them to date ranges.
+    Works across languages by using:
+    1. Universal date patterns (ISO dates, numeric formats)
+    2. Number-based relative time extraction
+    3. Common English patterns as fallback
+    """
+
+    def __init__(self):
+        # Universal date patterns (work across all languages)
+        # Order matters - more specific patterns first!
+        self.universal_patterns = {
+            # ISO date format: 2024-01-15, 2024-01-15T10:30:00 (HIGHEST PRIORITY)
+            r'(\d{4})-(\d{2})-(\d{2})(?:T[\d:]+)?': self._parse_iso_date,
+            
+            # Numeric date formats: 15/01/2024, 01/15/2024, 15.01.2024 (BEFORE year-only!)
+            r'\b(\d{1,2})[/\.\-](\d{1,2})[/\.\-](\d{4})\b': self._parse_numeric_date,
+            
+            # Month-year: 01/2024, 01-2024
+            r'\b(\d{1,2})[/\-](\d{4})\b': self._parse_month_year,
+            
+            # Year-month: 2024/01, 2024-01
+            r'\b(\d{4})[/\-](\d{1,2})\b': self._parse_month_year_reverse,
+            
+            # Year only: 2024, 2023 (LOWEST PRIORITY - catches lone years)
+            r'\b(20\d{2})\b': self._parse_year_only,
+        }
+        
+        # Pattern to extract numbers for relative time (language-agnostic)
+        # Examples: "7 días", "30 jours", "14 Tage", "últimos 7", "derniers 30"
+        # Matches: number + optional space + word OR word + optional space + number
+        self.relative_number_pattern = r'(\d+)\s*\w+|\w+\s+(\d+)'
+        
+        # English patterns for backward compatibility
+        self.english_patterns = {
```

Comment:
The naming seems incorrect, it sounds more something like `common_languages_patterns`?